File: Módulo5 - WebServices/WS3/ws provider/provider.py
# Importa a classe BaseHTTPRequestHandler para criar um handler HTTP personalizado
from http.server import BaseHTTPRequestHandler, HTTPServer
# Importa o módulo json para serializar e deserializar dados JSON
import json

# Define a função para validar um CPF (Cadastro de Pessoa Física)
def validar_cpf(cpf):
    # Remove todos os caracteres não numéricos do CPF, mantendo apenas dígitos
    cpf = ''.join(filter(str.isdigit, cpf))
    # Verifica se o CPF tem exatamente 11 dígitos após a limpeza
    if len(cpf) != 11:
        # Retorna False se o CPF não tiver 11 dígitos
        return False
    # Verifica se todos os dígitos são iguais (ex: 11111111111, 22222222222)
    if cpf == cpf[0] * 11:
        # Retorna False se todos os dígitos forem iguais (CPF inválido)
        return False

    # Define função interna para calcular um dígito verificador do CPF
    def calc_digito(cpf, peso):
        # Calcula a soma ponderada: multiplica cada dígito pela posição (peso - índice)
        # O peso começa em 10 para o primeiro dígito verificador e 11 para o segundo
        s = sum(int(cpf[i]) * (peso - i) for i in range(peso - 1))
        # Calcula o resto da divisão por 11 e multiplica por 10
        resto = (s * 10) % 11
        # Se o resto for 10, o dígito verificador é 0, caso contrário é o próprio resto
        return 0 if resto == 10 else resto

    # Calcula o primeiro dígito verificador (posição 9 do CPF) usando peso 10
    dig1 = calc_digito(cpf, 10)
    # Calcula o segundo dígito verificador (posição 10 do CPF) usando peso 11
    dig2 = calc_digito(cpf, 11)

    # Compara os dois últimos dígitos do CPF com os dígitos calculados
    # Retorna True se forem iguais (CPF válido), False caso contrário
    return cpf[-2:] == f"{dig1}{dig2}"


# Define a classe CPFService que herda de BaseHTTPRequestHandler
# Esta classe processa requisições HTTP para validação de CPF
class CPFService(BaseHTTPRequestHandler):

    # ------------------------------
    # GET /cpf?numero=XXXXXXXXXXX
    # ------------------------------
    # Método que processa requisições HTTP GET
    def do_GET(self):
        # Importa funções para fazer parse da URL e dos parâmetros de query
        from urllib.parse import urlparse, parse_qs
        # Faz o parse da URL da requisição para extrair path e query string
        parsed = urlparse(self.path)

        # Verifica se o caminho da requisição é "/cpf"
        if parsed.path == "/cpf":
            # Extrai os parâmetros da query string (ex: ?numero=11144477735)
            params = parse_qs(parsed.query)
            # Obtém o valor do parâmetro "numero", retornando string vazia se não existir
            # parse_qs retorna uma lista, então pegamos o primeiro elemento [0]
            cpf = params.get("numero", [""])[0]

            # Chama a função validar_cpf para verificar se o CPF é válido
            valido = validar_cpf(cpf)
            # Envia resposta JSON com status 200 (OK), contendo o CPF e se é válido
            self._send_json(200, {"cpf": cpf, "valido": valido})
            # Retorna para encerrar o processamento da requisição
            return

        # Se o caminho não for "/cpf", retorna erro 404 (não encontrado)
        self._send_json(404, {"erro": "rota não encontrada"})

    # ------------------------------
    # POST /cpf
    # Body: { "cpf": "11144477735" }
    # ------------------------------
    # Método que processa requisições HTTP POST
    def do_POST(self):
        # Verifica se o caminho da requisição é "/cpf"
        if self.path == "/cpf":
            # Obtém o tamanho do corpo da requisição do header Content-Length
            # Se não existir, assume 0
            tamanho = int(self.headers.get('Content-Length', 0))
            # Lê o corpo da requisição (bytes) e decodifica para string UTF-8
            corpo = self.rfile.read(tamanho).decode("utf-8")

            # Tenta fazer o parse do JSON do corpo da requisição
            try:
                # Converte a string JSON em um dicionário Python
                dados = json.loads(corpo)
                # Obtém o valor da chave "cpf" do JSON, retornando string vazia se não existir
                cpf = dados.get("cpf", "")
            except:
                # Se houver erro ao fazer parse do JSON, retorna erro 400 (Bad Request)
                self._send_json(400, {"erro": "JSON inválido"})
                # Retorna para encerrar o processamento
                return

            # Chama a função validar_cpf para verificar se o CPF é válido
            valido = validar_cpf(cpf)
            # Envia resposta JSON com status 200 (OK), contendo o CPF e se é válido
            self._send_json(200, {"cpf": cpf, "valido": valido})
            # Retorna para encerrar o processamento da requisição
            return

        # Se o caminho não for "/cpf", retorna erro 404 (não encontrado)
        self._send_json(404, {"erro": "rota não encontrada"})

    # ------------------------------
    # Helper para JSON
    # ------------------------------
    # Método auxiliar privado para enviar respostas JSON
    def _send_json(self, status, data):
        # Converte o dicionário Python em string JSON e depois em bytes
        resposta = json.dumps(data).encode()

        # Envia o código de status HTTP (ex: 200, 404, 400)
        self.send_response(status)
        # Define o header Content-Type como application/json
        self.send_header("Content-Type", "application/json")
        # Define o header Content-Length com o tamanho da resposta em bytes
        self.send_header("Content-Length", str(len(resposta)))
        # Finaliza o envio dos headers HTTP
        self.end_headers()
        # Escreve o corpo da resposta (JSON em bytes) no stream de saída
        self.wfile.write(resposta)


# Define a função run() que inicia o servidor HTTP
def run():
    # Cria uma instância de HTTPServer escutando em localhost na porta 8000
    # CPFService é a classe handler que processará as requisições
    servidor = HTTPServer(("localhost", 8000), CPFService)
    # Imprime mensagem informando que o servidor está rodando
    print("Servidor rodando em http://localhost:8000 ...")
    # Inicia o servidor e fica escutando requisições indefinidamente
    servidor.serve_forever()


# Verifica se o script está sendo executado diretamente (não importado como módulo)
if __name__ == "__main__":
    # Chama a função run() para iniciar o servidor
    run()

File: Módulo5 - WebServices/WS4/ws provider/Provider.java
// Importa classes do pacote com.sun.net.httpserver para criar servidor HTTP
import com.sun.net.httpserver.*;
// Importa classes de I/O do Java (InputStreamReader, BufferedReader, OutputStream, etc.)
import java.io.*;
// Importa InetSocketAddress para especificar endereço e porta do servidor
import java.net.InetSocketAddress;
// Importa Collectors para operações de stream (juntar linhas)
import java.util.stream.Collectors;

// Classe principal do provider de validação de CPF
public class Provider {

    // Método estático para validar CPF usando o algoritmo oficial da Receita Federal
    public static boolean validaCPF(String cpf) {
        // Remove todos os caracteres que não são dígitos usando regex (\\D significa não-dígito)
        cpf = cpf.replaceAll("\\D", "");
        // Verifica se o CPF não tem 11 dígitos OU se todos os caracteres são iguais (distinct().count() == 1)
        if (cpf.length() != 11 || cpf.chars().distinct().count() == 1)
            return false;

        // Loop para calcular os dois dígitos verificadores (posições 9 e 10, índices 9 e 10)
        for (int t = 9; t < 11; t++) {
            // Inicializa a variável soma com zero
            int sum = 0;
            // Itera sobre os dígitos do CPF até a posição t
            for (int i = 0; i < t; i++) {
                // Soma o produto do dígito na posição i (convertido para int) pelo peso ((t + 1) - i)
                sum += Character.getNumericValue(cpf.charAt(i)) * ((t + 1) - i);
            }
            // Calcula o dígito verificador: resto da divisão por 11 multiplicado por 10, depois resto por 10
            int dig = ((sum * 10) % 11) % 10;
            // Se o dígito calculado não corresponder ao dígito na posição t do CPF, retorna false
            if (dig != Character.getNumericValue(cpf.charAt(t)))
                return false;
        }
        // Se passou por todas as validações (tamanho, dígitos diferentes, dígitos verificadores corretos), o CPF é válido
        return true;
    }

    // Método principal que inicia o servidor HTTP
    // args: argumentos da linha de comando (não utilizados neste caso)
    // throws IOException: pode lançar exceção de I/O ao criar o servidor
    public static void main(String[] args) throws IOException {
        // Cria um servidor HTTP na porta 8082 (0 significa usar backlog padrão)
        HttpServer server = HttpServer.create(new InetSocketAddress(8082), 0);
        // Cria um contexto para a rota /cpf e associa ao handler MyHandler
        server.createContext("/cpf", new MyHandler());
        // Inicia o servidor (cria threads para processar requisições)
        server.start();
        // Imprime mensagem informando que o servidor está rodando
        System.out.println("Java CPF Provider rodando em http://localhost:8082");
    }

    // Classe interna que implementa HttpHandler para processar requisições HTTP
    static class MyHandler implements HttpHandler {
        // Método obrigatório da interface HttpHandler que processa cada requisição
        // ex: objeto HttpExchange que contém informações da requisição e permite enviar resposta
        // throws IOException: pode lançar exceção de I/O ao processar a requisição
        @Override
        public void handle(HttpExchange ex) throws IOException {
            // Obtém o método HTTP da requisição (GET, POST, etc.)
            String method = ex.getRequestMethod();
            // Inicializa a string de resposta vazia
            String response = "";

            // Verifica se o método HTTP é GET
            if (method.equals("GET")) {
                // Obtém a query string da URI da requisição
                String query = ex.getRequestURI().getQuery();
                // Verifica se a query string não é nula e contém o parâmetro "cpf="
                if (query != null && query.contains("cpf=")) {
                    // Extrai o valor do CPF da query string (substring a partir de "cpf=" + 4 caracteres)
                    String cpf = query.substring(query.indexOf("cpf=") + 4);
                    // Remove outros parâmetros se houver (se contém "&", pega apenas até o "&")
                    if (cpf.contains("&")) {
                        cpf = cpf.substring(0, cpf.indexOf("&"));
                    }
                    // Valida o CPF usando o método validaCPF
                    boolean valid = validaCPF(cpf);
                    // Monta a resposta JSON como string manualmente (concatenação de strings)
                    // valid é boolean, então será convertido para "true" ou "false" na string
                    response = "{\"cpf\":\"" + cpf + "\", \"valid\":" + valid + "}";
                } else {
                    // Se o parâmetro cpf não foi encontrado, monta resposta de erro
                    response = "{\"error\": \"parâmetro cpf não encontrado\"}";
                    // Adiciona header Content-Type como application/json
                    // getResponseHeaders() retorna um objeto Headers para manipular headers HTTP
                    ex.getResponseHeaders().add("Content-Type", "application/json");
                    // Envia resposta com status 400 (Bad Request) e tamanho da resposta em bytes
                    // sendResponseHeaders() envia os headers HTTP, incluindo status code
                    ex.sendResponseHeaders(400, response.length());
                    // Escreve os bytes da resposta no corpo da resposta
                    // getBytes() converte a string para array de bytes usando encoding padrão
                    ex.getResponseBody().write(response.getBytes());
                    // Fecha a conexão HTTP
                    ex.close();
                    // Retorna para evitar processar mais código
                    return;
                }
            } 
            // Verifica se o método HTTP é POST
            else if (method.equals("POST")) {
                // Cria um InputStreamReader para ler o corpo da requisição com encoding UTF-8
                InputStreamReader isr = new InputStreamReader(ex.getRequestBody(), "utf-8");
                // Lê todas as linhas do corpo da requisição e junta em uma única string
                String body = new BufferedReader(isr).lines().collect(Collectors.joining());
                // Tenta fazer parse do JSON do corpo da requisição
                try {
                    // Parse JSON simples: {"cpf": "11144477735"}
                    // Inicializa string vazia para o CPF
                    String cpf = "";
                    // Remove espaços em branco do início e fim do body
                    body = body.trim();
                    // Verifica se o body começa com "{" e contém a chave "cpf"
                    if (body.startsWith("{") && body.contains("\"cpf\"")) {
                        // Encontra a posição da string "cpf" no body
                        int cpfIndex = body.indexOf("\"cpf\"");
                        // Encontra a posição do ":" após "cpf" e avança 1 posição
                        int valueStart = body.indexOf(":", cpfIndex) + 1;
                        // Pula espaços e aspas
                        // Enquanto não chegou ao fim e o caractere atual é espaço ou aspas, avança
                        while (valueStart < body.length() && 
                               (body.charAt(valueStart) == ' ' || body.charAt(valueStart) == '"')) {
                            valueStart++;
                        }
                        // Inicializa o índice do fim do valor
                        int valueEnd = valueStart;
                        // Enquanto não chegou ao fim e o caractere atual não é aspas, vírgula ou chave de fechamento
                        while (valueEnd < body.length() && 
                               body.charAt(valueEnd) != '"' && 
                               body.charAt(valueEnd) != ',' && 
                               body.charAt(valueEnd) != '}') {
                            valueEnd++;
                        }
                        // Extrai o CPF como substring entre valueStart e valueEnd, removendo espaços
                        cpf = body.substring(valueStart, valueEnd).trim();
                    }
                    // Se o CPF estiver vazio após o parse, lança exceção
                    if (cpf.isEmpty()) {
                        throw new Exception("CPF não encontrado no JSON");
                    }
                    // Valida o CPF usando o método validaCPF
                    boolean valid = validaCPF(cpf);
                    // Monta a resposta JSON como string
                    response = "{\"cpf\":\"" + cpf + "\", \"valid\":" + valid + "}";
                // Se houver erro ao fazer parse do JSON
                } catch (Exception e) {
                    // Monta resposta de erro JSON
                    response = "{\"error\": \"JSON inválido\"}";
                    // Adiciona header Content-Type como application/json
                    ex.getResponseHeaders().add("Content-Type", "application/json");
                    // Envia resposta com status 400 (Bad Request) e tamanho da resposta
                    ex.sendResponseHeaders(400, response.length());
                    // Escreve os bytes da resposta no corpo da resposta
                    ex.getResponseBody().write(response.getBytes());
                    // Fecha a conexão
                    ex.close();
                    return;
                }
            } 
            // Se o método HTTP não for GET nem POST (PUT, DELETE, PATCH, etc.)
            else {
                // Monta resposta de erro indicando método não permitido
                response = "{\"error\": \"method not allowed\"}";
                // Envia resposta com status 405 (Method Not Allowed) e tamanho da resposta
                // 405 indica que o método HTTP usado não é permitido para este endpoint
                ex.sendResponseHeaders(405, response.length());
                // Escreve os bytes da resposta no corpo da resposta
                ex.getResponseBody().write(response.getBytes());
                // Fecha a conexão HTTP
                ex.close();
                // Retorna para evitar processar mais código
                return;
            }

            // Se chegou aqui, a requisição foi processada com sucesso (GET ou POST válido)
            // Adiciona header Content-Type como application/json
            ex.getResponseHeaders().add("Content-Type", "application/json");
            // Envia resposta com status 200 (OK) e tamanho da resposta em bytes
            // 200 indica que a requisição foi processada com sucesso
            ex.sendResponseHeaders(200, response.length());
            // Obtém o OutputStream para escrever a resposta
            // getResponseBody() retorna um OutputStream para escrever o corpo da resposta
            OutputStream os = ex.getResponseBody();
            // Escreve os bytes da resposta no corpo da resposta
            os.write(response.getBytes());
            // Fecha o OutputStream (também fecha a conexão HTTP)
            os.close();
        }
    }
}
